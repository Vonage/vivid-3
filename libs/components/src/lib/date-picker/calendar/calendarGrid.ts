import { addDays } from 'date-fns';
import type { DatePickerLocale } from '../locale';
import { type DateStr, formatDateStr } from './dateStr';

import { type Month } from './month';

export interface Weekday {
	name: string;
	shortName: string;
}

const getWeekdays = (locale: DatePickerLocale) => {
	const firstDayOfWeek = locale.firstDayOfWeek;
	const localeWeekdayToNative = (day: number): number =>
		(day + firstDayOfWeek) % 7;

	const days = [];
	for (let i = 0; i < 7; i++) {
		days.push({
			name: locale.weekdays.name[localeWeekdayToNative(i)],
			shortName: locale.weekdays.shorthand[localeWeekdayToNative(i)],
		});
	}
	return days;
};

export interface CalendarGridDate {
	date: DateStr;
	label: string;
	isOutsideMonth: boolean;
}

const gridDate = (date: Date, isOutsideMonth: boolean): CalendarGridDate => ({
	date: formatDateStr(date),
	label: `${date.getDate()}`,
	isOutsideMonth,
});

export type CalendarGrid = {
	weekdays: Weekday[];
	grid: CalendarGridDate[][];
};

const buildDateGrid = (
	{ month, year }: Month,
	getDay: (date: Date) => number
) => {
	const grid: CalendarGridDate[][] = [];

	const firstDay = new Date(year, month, 1);
	const lastDay = new Date(year, month + 1, 0);
	const daysInMonth = lastDay.getDate();
	const firstDayInWeek = getDay(firstDay);

	let week: CalendarGridDate[] = [];

	// Fill in the days before the first day of the month
	for (let i = 0; i < firstDayInWeek; i++) {
		week.push(gridDate(addDays(firstDay, i - firstDayInWeek), true));
	}

	// Fill up days of the month
	for (let i = 1; i <= daysInMonth; i++) {
		week.push(gridDate(new Date(year, month, i), false));
		if (week.length === 7) {
			grid.push(week);
			week = [];
		}
	}

	// Fill in the days after the last day of the month
	const daysInLastWeek = week.length;
	for (let i = daysInLastWeek; i < 7; i++) {
		week.push(gridDate(addDays(lastDay, i - daysInLastWeek + 1), true));
	}

	if (week.length > 0) {
		grid.push(week);
	}

	return grid;
};

export const buildCalendarGrid = (
	month: Month,
	locale: DatePickerLocale
): CalendarGrid => {
	// Shift week days to start from firstDayOfWeek
	const { firstDayOfWeek } = locale;
	const getShiftedDay = (date: Date): number =>
		(date.getDay() - firstDayOfWeek + 7) % 7;

	return {
		weekdays: getWeekdays(locale),
		grid: buildDateGrid(month, getShiftedDay),
	};
};
