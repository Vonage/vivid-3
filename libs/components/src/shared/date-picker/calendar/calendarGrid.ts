import { addDays } from 'date-fns';
import type { DatePickerLocale } from '../locale';
import { type DateStr, formatDateStr } from './dateStr';

import { type Month } from './month';

export interface Weekday {
	name: string;
	shortName: string;
}

const getWeekdays = (locale: DatePickerLocale) => {
	const firstDayOfWeek = locale.firstDayOfWeek;
	const localeWeekdayToNative = (day: number): number =>
		(day + firstDayOfWeek) % 7;

	const days = [];
	for (let i = 0; i < 7; i++) {
		days.push({
			name: locale.weekdays.name[localeWeekdayToNative(i)],
			shortName: locale.weekdays.shorthand[localeWeekdayToNative(i)],
		});
	}
	return days;
};

export interface CalendarGridDate {
	date: DateStr;
	label: string;
	isOutsideMonth: boolean;
}

const gridDate = (date: Date, isOutsideMonth: boolean): CalendarGridDate => ({
	date: formatDateStr(date),
	label: `${date.getDate()}`,
	isOutsideMonth,
});

export type CalendarGrid = {
	weekdays: Weekday[];
	grid: CalendarGridDate[][];
};

function isOutsideMonth(dayIndexInMonth: number, daysInMonth: number) {
	return dayIndexInMonth < 0 || dayIndexInMonth >= daysInMonth;
}

const buildDateGrid = (
	{ month, year }: Month,
	getDay: (date: Date) => number
) => {
	const grid: CalendarGridDate[][] = [];

	const firstDay = new Date(year, month, 1);
	const lastDay = new Date(year, month + 1, 0);
	const firstDayInWeek = getDay(firstDay);
	const daysInMonth = lastDay.getDate();
	const daysOutsideMonthInLastWeek = 7 - getDay(lastDay);
	const totalDaysInCalendar =
		daysInMonth + firstDayInWeek + daysOutsideMonthInLastWeek;
	let week: CalendarGridDate[] = [];

	for (let i = 0; i < totalDaysInCalendar; i++) {
		const dayIndexInMonth = i - firstDayInWeek;
		week.push(
			gridDate(
				addDays(firstDay, dayIndexInMonth),
				isOutsideMonth(dayIndexInMonth, daysInMonth)
			)
		);
		if (week.length === 7) {
			grid.push(week);
			week = [];
		}
	}

	return grid;
};

export const buildCalendarGrid = (
	month: Month,
	locale: DatePickerLocale
): CalendarGrid => {
	// Shift week days to start from firstDayOfWeek
	const { firstDayOfWeek } = locale;
	const getShiftedDay = (date: Date): number =>
		(date.getDay() - firstDayOfWeek + 7) % 7;

	return {
		weekdays: getWeekdays(locale),
		grid: buildDateGrid(month, getShiftedDay),
	};
};
