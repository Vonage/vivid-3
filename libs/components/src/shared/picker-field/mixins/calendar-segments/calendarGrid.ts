import { addDays } from 'date-fns';
import type { CalendarPickerLocale } from '../calendar-picker.locale';
import { type DateStr, formatDateStr } from '../../../datetime/dateStr';

import { type Month } from '../../../datetime/month';

export interface Weekday {
	name: string;
	shortName: string;
}

const getWeekdays = (locale: CalendarPickerLocale) => {
	const firstDayOfWeek = locale.firstDayOfWeek;
	const localeWeekdayToNative = (day: number): number =>
		(day + firstDayOfWeek) % 7;

	const days = [];
	for (let i = 0; i < 7; i++) {
		days.push({
			name: locale.weekdays.name[localeWeekdayToNative(i)],
			shortName: locale.weekdays.shorthand[localeWeekdayToNative(i)],
		});
	}
	return days;
};

export interface CalendarGridDate {
	date: DateStr;
	label: string;
	isOutsideMonth: boolean;
}

const gridDate = (date: Date, isOutsideMonth: boolean): CalendarGridDate => ({
	date: formatDateStr(date),
	label: `${date.getDate()}`,
	isOutsideMonth,
});

export type CalendarGrid = {
	weekdays: Weekday[];
	grid: CalendarGridDate[][];
};

function isOutsideMonth(dayIndexInMonth: number, daysInMonth: number) {
	return dayIndexInMonth < 0 || dayIndexInMonth >= daysInMonth;
}

const buildDateGrid = (
	{ month, year }: Month,
	getDay: (date: Date) => number
) => {
	const grid: CalendarGridDate[][] = [];

	const firstDay = new Date(year, month, 1);
	const lastDay = new Date(year, month + 1, 0);
	const firstDayInWeek = getDay(firstDay);
	const daysInMonth = lastDay.getDate();
	const daysOutsideMonthInLastWeek = 7 - getDay(lastDay);
	const totalDaysInCalendar =
		daysInMonth + firstDayInWeek + daysOutsideMonthInLastWeek;
	let week: CalendarGridDate[] = [];

	for (let i = 0; i < totalDaysInCalendar; i++) {
		const dayIndexInMonth = i - firstDayInWeek;
		week.push(
			gridDate(
				addDays(firstDay, dayIndexInMonth),
				isOutsideMonth(dayIndexInMonth, daysInMonth)
			)
		);
		if (week.length === 7) {
			grid.push(week);
			week = [];
		}
	}

	return grid;
};

export const buildCalendarGrid = (
	month: Month,
	locale: CalendarPickerLocale
): CalendarGrid => {
	// Shift week days to start from firstDayOfWeek
	const { firstDayOfWeek } = locale;
	const getShiftedDay = (date: Date): number =>
		(date.getDay() - firstDayOfWeek + 7) % 7;

	return {
		weekdays: getWeekdays(locale),
		grid: buildDateGrid(month, getShiftedDay),
	};
};
