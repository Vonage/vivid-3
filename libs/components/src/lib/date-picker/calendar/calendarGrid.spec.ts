import type { DatePickerLocale } from '../locale.ts';
import { buildCalendarGrid } from './calendarGrid.ts';

const enUsLocale = {
	weekdays: {
		name: [
			'Sunday',
			'Monday',
			'Tuesday',
			'Wednesday',
			'Thursday',
			'Friday',
			'Saturday',
		],
		shorthand: ['Sun', 'Mon', 'Tue', 'Wed', 'Thu', 'Fri', 'Sat'],
	},
	firstDayOfWeek: 0,
} as DatePickerLocale;

const enGbLocale = { ...enUsLocale, firstDayOfWeek: 1 };

describe('buildCalendarGrid', () => {
	it('should return weekdays with sunday first for US locale', function () {
		const expectedOutput = [
			{
				'name': 'Sunday',
				'shortName': 'Sun',
			},
			{
				'name': 'Monday',
				'shortName': 'Mon',
			},
			{
				'name': 'Tuesday',
				'shortName': 'Tue',
			},
			{
				'name': 'Wednesday',
				'shortName': 'Wed',
			},
			{
				'name': 'Thursday',
				'shortName': 'Thu',
			},
			{
				'name': 'Friday',
				'shortName': 'Fri',
			},
			{
				'name': 'Saturday',
				'shortName': 'Sat',
			},
		];
		const grid = buildCalendarGrid({ month: 7, year: 2023 }, enUsLocale);
		expect(grid.weekdays).toEqual(expectedOutput);
	});

	it('should return weekdays with monday first for GB locale', function () {
		const expectedOutput = [
			{
				'name': 'Monday',
				'shortName': 'Mon',
			},
			{
				'name': 'Tuesday',
				'shortName': 'Tue',
			},
			{
				'name': 'Wednesday',
				'shortName': 'Wed',
			},
			{
				'name': 'Thursday',
				'shortName': 'Thu',
			},
			{
				'name': 'Friday',
				'shortName': 'Fri',
			},
			{
				'name': 'Saturday',
				'shortName': 'Sat',
			},
			{
				'name': 'Sunday',
				'shortName': 'Sun',
			},
		];
		const grid = buildCalendarGrid({ month: 7, year: 2023 }, enGbLocale);
		expect(grid.weekdays).toEqual(expectedOutput);
	});

	it('should return the first week of the month', function () {
		const monthWithFirstDayOnSunday = { year: 2023, month: 9};
		const grid = buildCalendarGrid(monthWithFirstDayOnSunday, enUsLocale);
		const isFirstWeekInMonth = grid.grid[0]
			.every((day, index) => !day.isOutsideMonth && day.label === (index + 1).toString());
		expect(isFirstWeekInMonth).toBe(true);
	});

	it('should return the first week of the month starting on tuesday with two days from the previous month', function () {
		const expectedOutput = [
			{ label: '30', isOutsideMonth: true, date: '2023-07-30' },
			{ label: '31', isOutsideMonth: true, date: '2023-07-31' },
			{ label: '1', isOutsideMonth: false, date: '2023-08-01' },
			{ label: '2', isOutsideMonth: false, date: '2023-08-02' },
			{ label: '3', isOutsideMonth: false, date: '2023-08-03' },
			{ label: '4', isOutsideMonth: false, date: '2023-08-04' },
			{ label: '5', isOutsideMonth: false, date: '2023-08-05' },
		];
		const monthWithFirstDayOnTuesday = { year: 2023, month: 7};
		const grid = buildCalendarGrid(monthWithFirstDayOnTuesday, enUsLocale);
		expect(grid.grid[0]).toEqual(expectedOutput);
	});

	it('should return the last week of the month sunday to saturday', function () {
		const monthWithLastDayOnSaturday = { year: 2023, month: 8};
		const grid = buildCalendarGrid(monthWithLastDayOnSaturday, enUsLocale);
		const isLastWeekInMonth = grid.grid[4]
			.every((day, index) => !day.isOutsideMonth && day.label === (index + 30 - 6).toString());
		expect(isLastWeekInMonth).toBe(true);
	});

	it('should return the last week of the month padded with next month first days', function () {
		const expectedOutput = [
			{ label: '29', isOutsideMonth: false, date: '2023-10-29' },
			{ label: '30', isOutsideMonth: false, date: '2023-10-30' },
			{ label: '31', isOutsideMonth: false, date: '2023-10-31' },
			{ label: '1', isOutsideMonth: true, date: '2023-11-01' },
			{ label: '2', isOutsideMonth: true, date: '2023-11-02' },
			{ label: '3', isOutsideMonth: true, date: '2023-11-03' },
			{ label: '4', isOutsideMonth: true, date: '2023-11-04' },
		];
		const monthWithLastDayOnSaturday = { year: 2023, month: 9};
		const grid = buildCalendarGrid(monthWithLastDayOnSaturday, enUsLocale);
		expect(grid.grid[4]).toEqual(expectedOutput);
	});

	it('should include a cell for dates within the month', () => {
		const grid = buildCalendarGrid({ month: 7, year: 2023 }, enUsLocale);
		const thirdDayOfFirstWeek = grid.grid[0][2];
		expect(thirdDayOfFirstWeek.date).toBe('2023-08-01');
		expect(thirdDayOfFirstWeek.label).toBe('1');
		expect(thirdDayOfFirstWeek.isOutsideMonth).toBe(false);
	});

	it('should include a cell for dates in the week before the first of the month', () => {
		const grid = buildCalendarGrid({ month: 7, year: 2023 }, enUsLocale);
		const firstDayOfFirstWeek = grid.grid[0][0];
		expect(firstDayOfFirstWeek.date).toBe('2023-07-30');
		expect(firstDayOfFirstWeek.label).toBe('30');
		expect(firstDayOfFirstWeek.isOutsideMonth).toBe(true);
	});

	it('should include a cell for dates in the week after the last of the month', () => {
		const grid = buildCalendarGrid({ month: 7, year: 2023 }, enUsLocale);
		const lastDayOfLastWeek = grid.grid[4][6];
		expect(lastDayOfLastWeek.date).toBe('2023-09-02');
		expect(lastDayOfLastWeek.label).toBe('2');
		expect(lastDayOfLastWeek.isOutsideMonth).toBe(true);
	});

	it('should return grid for US locale', function () {
		const grid = buildCalendarGrid({ month: 7, year: 2022 }, enUsLocale);
		expect(grid.grid).toMatchSnapshot();
	});

	it('should return grid for GB locale', function () {
		const grid = buildCalendarGrid({ month: 7, year: 2024 }, enGbLocale);
		expect(grid.grid).toMatchSnapshot();
	});
});
