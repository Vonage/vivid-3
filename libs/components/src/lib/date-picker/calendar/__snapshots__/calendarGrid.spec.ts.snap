// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`buildCalendarGrid should return grid for GB locale 1`] = `
Array [
  Array [
    Object {
      "date": "2024-07-29",
      "isOutsideMonth": true,
      "label": "29",
    },
    Object {
      "date": "2024-07-30",
      "isOutsideMonth": true,
      "label": "30",
    },
    Object {
      "date": "2024-07-31",
      "isOutsideMonth": true,
      "label": "31",
    },
    Object {
      "date": "2024-08-01",
      "isOutsideMonth": false,
      "label": "1",
    },
    Object {
      "date": "2024-08-02",
      "isOutsideMonth": false,
      "label": "2",
    },
    Object {
      "date": "2024-08-03",
      "isOutsideMonth": false,
      "label": "3",
    },
    Object {
      "date": "2024-08-04",
      "isOutsideMonth": false,
      "label": "4",
    },
  ],
  Array [
    Object {
      "date": "2024-08-05",
      "isOutsideMonth": false,
      "label": "5",
    },
    Object {
      "date": "2024-08-06",
      "isOutsideMonth": false,
      "label": "6",
    },
    Object {
      "date": "2024-08-07",
      "isOutsideMonth": false,
      "label": "7",
    },
    Object {
      "date": "2024-08-08",
      "isOutsideMonth": false,
      "label": "8",
    },
    Object {
      "date": "2024-08-09",
      "isOutsideMonth": false,
      "label": "9",
    },
    Object {
      "date": "2024-08-10",
      "isOutsideMonth": false,
      "label": "10",
    },
    Object {
      "date": "2024-08-11",
      "isOutsideMonth": false,
      "label": "11",
    },
  ],
  Array [
    Object {
      "date": "2024-08-12",
      "isOutsideMonth": false,
      "label": "12",
    },
    Object {
      "date": "2024-08-13",
      "isOutsideMonth": false,
      "label": "13",
    },
    Object {
      "date": "2024-08-14",
      "isOutsideMonth": false,
      "label": "14",
    },
    Object {
      "date": "2024-08-15",
      "isOutsideMonth": false,
      "label": "15",
    },
    Object {
      "date": "2024-08-16",
      "isOutsideMonth": false,
      "label": "16",
    },
    Object {
      "date": "2024-08-17",
      "isOutsideMonth": false,
      "label": "17",
    },
    Object {
      "date": "2024-08-18",
      "isOutsideMonth": false,
      "label": "18",
    },
  ],
  Array [
    Object {
      "date": "2024-08-19",
      "isOutsideMonth": false,
      "label": "19",
    },
    Object {
      "date": "2024-08-20",
      "isOutsideMonth": false,
      "label": "20",
    },
    Object {
      "date": "2024-08-21",
      "isOutsideMonth": false,
      "label": "21",
    },
    Object {
      "date": "2024-08-22",
      "isOutsideMonth": false,
      "label": "22",
    },
    Object {
      "date": "2024-08-23",
      "isOutsideMonth": false,
      "label": "23",
    },
    Object {
      "date": "2024-08-24",
      "isOutsideMonth": false,
      "label": "24",
    },
    Object {
      "date": "2024-08-25",
      "isOutsideMonth": false,
      "label": "25",
    },
  ],
  Array [
    Object {
      "date": "2024-08-26",
      "isOutsideMonth": false,
      "label": "26",
    },
    Object {
      "date": "2024-08-27",
      "isOutsideMonth": false,
      "label": "27",
    },
    Object {
      "date": "2024-08-28",
      "isOutsideMonth": false,
      "label": "28",
    },
    Object {
      "date": "2024-08-29",
      "isOutsideMonth": false,
      "label": "29",
    },
    Object {
      "date": "2024-08-30",
      "isOutsideMonth": false,
      "label": "30",
    },
    Object {
      "date": "2024-08-31",
      "isOutsideMonth": false,
      "label": "31",
    },
    Object {
      "date": "2024-09-01",
      "isOutsideMonth": true,
      "label": "1",
    },
  ],
]
`;

exports[`buildCalendarGrid should return grid for US locale 1`] = `
Array [
  Array [
    Object {
      "date": "2022-07-31",
      "isOutsideMonth": true,
      "label": "31",
    },
    Object {
      "date": "2022-08-01",
      "isOutsideMonth": false,
      "label": "1",
    },
    Object {
      "date": "2022-08-02",
      "isOutsideMonth": false,
      "label": "2",
    },
    Object {
      "date": "2022-08-03",
      "isOutsideMonth": false,
      "label": "3",
    },
    Object {
      "date": "2022-08-04",
      "isOutsideMonth": false,
      "label": "4",
    },
    Object {
      "date": "2022-08-05",
      "isOutsideMonth": false,
      "label": "5",
    },
    Object {
      "date": "2022-08-06",
      "isOutsideMonth": false,
      "label": "6",
    },
  ],
  Array [
    Object {
      "date": "2022-08-07",
      "isOutsideMonth": false,
      "label": "7",
    },
    Object {
      "date": "2022-08-08",
      "isOutsideMonth": false,
      "label": "8",
    },
    Object {
      "date": "2022-08-09",
      "isOutsideMonth": false,
      "label": "9",
    },
    Object {
      "date": "2022-08-10",
      "isOutsideMonth": false,
      "label": "10",
    },
    Object {
      "date": "2022-08-11",
      "isOutsideMonth": false,
      "label": "11",
    },
    Object {
      "date": "2022-08-12",
      "isOutsideMonth": false,
      "label": "12",
    },
    Object {
      "date": "2022-08-13",
      "isOutsideMonth": false,
      "label": "13",
    },
  ],
  Array [
    Object {
      "date": "2022-08-14",
      "isOutsideMonth": false,
      "label": "14",
    },
    Object {
      "date": "2022-08-15",
      "isOutsideMonth": false,
      "label": "15",
    },
    Object {
      "date": "2022-08-16",
      "isOutsideMonth": false,
      "label": "16",
    },
    Object {
      "date": "2022-08-17",
      "isOutsideMonth": false,
      "label": "17",
    },
    Object {
      "date": "2022-08-18",
      "isOutsideMonth": false,
      "label": "18",
    },
    Object {
      "date": "2022-08-19",
      "isOutsideMonth": false,
      "label": "19",
    },
    Object {
      "date": "2022-08-20",
      "isOutsideMonth": false,
      "label": "20",
    },
  ],
  Array [
    Object {
      "date": "2022-08-21",
      "isOutsideMonth": false,
      "label": "21",
    },
    Object {
      "date": "2022-08-22",
      "isOutsideMonth": false,
      "label": "22",
    },
    Object {
      "date": "2022-08-23",
      "isOutsideMonth": false,
      "label": "23",
    },
    Object {
      "date": "2022-08-24",
      "isOutsideMonth": false,
      "label": "24",
    },
    Object {
      "date": "2022-08-25",
      "isOutsideMonth": false,
      "label": "25",
    },
    Object {
      "date": "2022-08-26",
      "isOutsideMonth": false,
      "label": "26",
    },
    Object {
      "date": "2022-08-27",
      "isOutsideMonth": false,
      "label": "27",
    },
  ],
  Array [
    Object {
      "date": "2022-08-28",
      "isOutsideMonth": false,
      "label": "28",
    },
    Object {
      "date": "2022-08-29",
      "isOutsideMonth": false,
      "label": "29",
    },
    Object {
      "date": "2022-08-30",
      "isOutsideMonth": false,
      "label": "30",
    },
    Object {
      "date": "2022-08-31",
      "isOutsideMonth": false,
      "label": "31",
    },
    Object {
      "date": "2022-09-01",
      "isOutsideMonth": true,
      "label": "1",
    },
    Object {
      "date": "2022-09-02",
      "isOutsideMonth": true,
      "label": "2",
    },
    Object {
      "date": "2022-09-03",
      "isOutsideMonth": true,
      "label": "3",
    },
  ],
]
`;
